src/waitFor.ts

```diff
@@ -196,7 +196,10 @@ export default async function waitFor<T>(
     setReactActEnvironment(false);
 
     try {
-      return await waitForInternal(expectation, optionsWithStackTrace);
+      const result = await waitForInternal(expectation, optionsWithStackTrace);
+      // drain the microtask queue before restoring the `act` environment
+      await new Promise((resolve) => setImmediate(resolve));
+      return result;
```

Comment:
I can see in the reference PR that there is
```ts
if (jestFakeTimersAreEnabled()) {
  jest.advanceTimersByTime(0)
}
```

Do you know why this is not needed here?

Did your added test failed with both fake and real timers?

Reply:
Oh, that's a really good catch, this `setImmediate` problem explains the issue we've had in #1347, it makes sense to me.